_drafts/practical_zephyr_01_kconfig.md

```diff
@@ -0,0 +1,986 @@
+
+![]({% img_url practical-zephyr/kconfig-menuconfig.png %})
+![]({% img_url practical-zephyr/kconfig-guiconfig.png %})
+
+For the sake of simplicity, we'll use `menuconfig` in the remainder of this article. In both tools, you can search for configuration symbols. In `menuconfig` - similar to `vim` - you can use the forward slash `/` to start searching, e.g., for our `PRINTK` symbol. Several results match, but only one of them is the plain symbol `PRINTK` which also shows the same description that we've seen in the `Kconfig` file:
+
+```
+PRINTK(=y) "Send printk() to console"
+```
+
+Navigate to the symbol either using the search or using the path _Subsystems and OS Services > Debugging Options_. You should see the following screen:
+
+![]({% img_url practical-zephyr/kconfig-menuconfig-debug.png %})
+
+Typically, you can toggle `bool` symbols using `Space` or `Enter`, but this does not seem to be possible for `PRINTK`. You can see that a symbol cannot be changed - which is indicated by the marker `*` being enclosed by dashes `-*-` instead of square brackets `[*]`. Using `?` we can display the symbol's information:
+
+```
+Name: PRINTK
+Prompt: Send printk() to console
+Type: bool
+Value: y
+
+Help:
+
+  This option directs printk() debugging output to the supported
+  console device, rather than suppressing the generation
+  of printk() output entirely. Output is sent immediately, without
+  any mutual exclusion or buffering.
+
+Default:
+  - y
+
+Symbols currently y-selecting this symbol:
+  - BOOT_BANNER
+...
+```
+
+We found the reason why we can't disable the symbol at the very bottom fo the `PRINTK`'s symbol information! The symbol `BOOT_BANNER` is `y-selecting` the `PRINTK` symbol, which is why we can't disable it and why the application's configuration has no effect. This simple example should also make you realize that it is hard if not impossible to figure out dependencies by navigating `Kconfig` fragments file by file. Let's fix our configuration:
+
+* Navigate to the `BOOT_BANNER` symbol and disable it.
+* Going back to `PRINTK` it should now be possible to disable it.
+
+
+### Persisting `menuconfig` and `guiconfig` changes
+
+`menuconfig` and `guiconfig` can be used to test and explore configuration options. Typically, you'd use the _Save_ operation to store the new configuration to the `zephyr/.config` file in your build directory. You can then also launch `menuconfig` multiple times and change the configuration to your needs, and the changes will always be reflected in the next build.
+
+The downside of saving the configuration to `zephyr/.config` in the build directory is that **all changes will be lost** once you perform a clean build, e.g., using `west build --pristine`. To persist your changes, you need to place them into your [configuration files](https://docs.zephyrproject.org/latest/build/kconfig/setting.html#setting-kconfig-configuration-values).
```

Comment:
To be honest I think the "Save" options in the tools are very misleading and it's way too easy to lose changes for my taste. "Patch" or "Fix" or anything other than "Save" would be a better take since "Save" for me still means that "you're changes are _saved_" - and not that they'll be lost with the next build :D